ipa-core/src/protocol/basics/reveal.rs

```diff
@@ -17,10 +17,117 @@ use crate::{
+    fn access_without_verification(self) -> T;
+}
+
+impl<T> ThisCodeIsAuthorizedToObserveRevealedValues<T> for UnauthorizedRevealWrapper<T> {
+    fn access_without_verification(self) -> T {
+        self.0
+    }
+}
+
+/// Raw semi-honest reveal protocol.
+///
+/// This should only be used in rare cases. Prefer to use the `Reveal` trait or the `reveal` function.
+///
+/// # Errors
+/// If there is a problem communicating with other helpers.
+pub async fn semi_honest_reveal<'fut, C, V, const N: usize>(
+    value: &'fut Replicated<V, N>,
+    ctx: C,
+    record_id: RecordId,
+    excluded: Option<Role>,
+) -> Result<Option<<V as Vectorizable<N>>::Array>, Error>
+where
+    C: Context + 'fut,
+    V: SharedValue + Vectorizable<N>,
+{
+    let left = value.left_arr();
+    let right = value.right_arr();
+
+    // Send shares, unless the target helper is excluded
+    if Some(ctx.role().peer(Direction::Right)) != excluded {
+        ctx.send_channel::<<V as Vectorizable<N>>::Array>(ctx.role().peer(Direction::Right))
+            .send(record_id, left)
+            .await?;
+    }
+
+    if Some(ctx.role()) == excluded {
+        Ok(None)
+    } else {
+        // Sleep until `helper's left` sends their share
+        let share: <V as Vectorizable<N>>::Array = ctx
+            .recv_channel(ctx.role().peer(Direction::Left))
+            .receive(record_id)
+            .await?;
+
+        Ok(Some(share + left + right))
+    }
+}
+
+async fn malicious_reveal<'fut, C, F>(
+    value: &'fut MaliciousReplicated<F>,
+    ctx: C,
+    record_id: RecordId,
+    excluded: Option<Role>,
+) -> Result<Option<<F as Vectorizable<1>>::Array>, Error>
+where
+    C: Context + 'fut,
+    F: ExtendableField,
+{
+    use futures::future::try_join;
+
+    use crate::secret_sharing::replicated::malicious::ThisCodeIsAuthorizedToDowngradeFromMalicious;
+
+    let (left, right) = value.x().access_without_downgrade().as_tuple();
+    let left_sender = ctx.send_channel(ctx.role().peer(Direction::Left));
+    let left_receiver = ctx.recv_channel::<F>(ctx.role().peer(Direction::Left));
+    let right_sender = ctx.send_channel(ctx.role().peer(Direction::Right));
+    let right_receiver = ctx.recv_channel::<F>(ctx.role().peer(Direction::Right));
+
+    // Send shares to the left and right helpers, unless excluded.
+    let send_left_fut =
+        MaybeFuture::future_or_ok(Some(ctx.role().peer(Direction::Left)) != excluded, || {
+            left_sender.send(record_id, right)
+        });
+
+    let send_right_fut =
+        MaybeFuture::future_or_ok(Some(ctx.role().peer(Direction::Right)) != excluded, || {
+            right_sender.send(record_id, left)
+        });
+    try_join(send_left_fut, send_right_fut).await?;
+
+    if Some(ctx.role()) == excluded {
+        Ok(None)
+    } else {
+        let (share_from_left, share_from_right) = try_join(
+            left_receiver.receive(record_id),
+            right_receiver.receive(record_id),
+        )
```

Comment:
Is it OK to await the `sends` first and only then await the `receives`? Wouldn't it be better instead to do all 4 in parallel with a try_join4?